rfc-drafts/yield_safe_lint.md

```diff
@@ -0,0 +1,89 @@
+
+# Guide-level explanation
+
+Provide a lint that can be attached to structs to let the compiler know that this struct is unsafe to be held across a yield boundary.
+
+```rust
+    #[yield_unsafe]
+    struct MyUnsafeYieldStruct {}
+```
+
+This struct if held across a yield boundary would cause a warning:
+
+```rust
+    async fn foo() {
+      let my_struct = MyUnsafeYieldStruct {};
+      my_async_op.await;
+      println!("{:?}", my_struct);
+    }
+```
+
+The compiler might output something along the lines of:
+
+
+    warning: Holding `MyUnsafeYieldStruct` across the await bound on line 3 might cause side effects.
+
+Examples use cases for this lint:
+
+    - `MutexGuard` holding this across a yield boundary in a single threaded executor could cause deadlocks. In a multi-threaded runtime the resulting future would become `!Send` which will stop the user from spawning this future and causing issues. But in a single threaded runtime which accepts `!Send` futures deadlocks could happen.
+        - The same applies to other such synchronization primitives such as locks from ParkingLot.
+    - `tracing::Span` has the ability to enter the span via the `tracing::span::Entered` guard. While entering a span is totally normal, during an async fn the span only needs to be entered during 
+
+This lint will enable the compiler to warn the user that the generated MIR could produce unforeseen side effects.
+An unsafe yield struct is some struct that may cause side effects when it is stored internally to the generator. Some examples of this are 
+
+This will be a best effort lint to signal to the user about unsafety of using certain types across yield points.
+
+(somewhere go into that async fn/generators generate new “code” that is different than what the users sees, this causes the problem)
+
+Signal incorrect usage of `!Send` types across yield points.
+
+Examples:
+
+- `[MutexGuard](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html)`
+- https://docs.rs/tracing/0.1.15/tracing/span/struct.Entered.html
+
+# Reference-level explanation
+
+**TODO**
+
+- [ ] Draw ideas from `#[must_use]` implementation
+ - [ ] Go through generator implementation to understand how values are captured in state machine.
+
+# Drawbacks
+- There is a possibility it can produce a false positive warning and it could get noisy. We likely want to allow overriding via some sort of module level `allow` attribute.
+    - This is probab
+
+
+# Rationale and alternatives
```

Comment:
I'm curious why it was chosen to use an attribute instead of a trait mirroring [`std::panic::UnwindSafe`](https://doc.rust-lang.org/std/panic/trait.UnwindSafe.html).

Reply:
I think @nikomatsakis wanted us to also take a look at what is going on with `MustUse`. I think we still want this to be some sort of attribute similar to `must_use` but we could also support a trait style too.

Reply:
I would want to parallel `#[must_use]`, I think? But I've shifted my opinion and am now open to using a trait for both I think.